Modular extensions for kube scheduler perf testing framework

ravisantoshgudimetla opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45973

**Special notes for your reviewer**:
It is not same as the existing one, the previous one has a single nodeaffinity key with multiple values. This one has multiple keys, values.

**Release note**:


<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```

ravisantoshgudimetla commented:
/cc @jayunit100 @gmarek @sjug

jayunit100 commented on test/integration/scheduler_perf/scheduler_perf_types.go:
```diff
@@ -18,11 +18,14 @@ package benchmark
 
 // High Level Configuration for all predicates and priorities.
 type schedulerPerfConfig struct {
+	NodeCount    int // It means if the testcase has 1000 nodes, out of them only nodeCount will have needed predicates or priorities.
```
I think you mean to say: 
```
NodeCount int // The number of nodes which will be seeded with metadata to match predicates and have non-trivial priority rankings

```

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -203,6 +116,8 @@ func schedulePods(config *testConfig) int32 {
 		// 30,000 pods -> wait till @ least 300 are scheduled to start measuring.
 		// TODO Find out why sometimes there may be scheduling blips in the beggining.
 		if len(scheduled) > config.numPods/100 {
+			//fmt.Println(config.numPods/100)
```
delete these lines

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -240,107 +155,133 @@ func schedulePods(config *testConfig) int32 {
-			Count:    numNodes / numGroups,
-			Strategy: testutils.NewLabelNodePrepareStrategy(nodeAffinityKey, fmt.Sprintf("%v", i)),
-		})
+// getBaseNodeTemplate returns the base node template.
```
should this just be a  const var ? I don't see any actual functional code in here  ?

ravisantoshgudimetla replied:
TBH, I am kind of against this to be used as constant because this base template could be configurable in future. Like CPU, memory etc. I can add a TODO fo this.(like user can provide the inputs or if clusterloader drives the whole thing, it can invoke this function with CPU, memory etc). But I am open to other options too.

sjug replied:
Yeah the idea is that you pass parameters to change the config, but you have none?

ravisantoshgudimetla replied:
As of now, @sjug but I will add a TODO, saying that, I have function in first place so as to make it configurable later. Do you want those parameters to be present now?

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -240,107 +155,133 @@ func schedulePods(config *testConfig) int32 {
-							},
-						},
+// getBasePodTemplate returns the base pod template.
+func getBasePodTemplate() *v1.Pod {
```
same comment as above, non functional, maybe just a const var  (or else, lets add some functionality in here if some is needed or planned for later)

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -240,107 +155,133 @@ func schedulePods(config *testConfig) int32 {
+}
+
+// mutateNode returns the modified node needed for creation of nodes.
+func (na nodeAffinity) mutateNode(node *v1.Node) {
```
makes sense... this is the plan for all pred/prios ? if it works that will be good i think

ravisantoshgudimetla replied:
Yes, every predicate and priority will have this. I tried with taints and it was able to pick up, I will add taints in another PR though.

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -240,107 +155,133 @@ func schedulePods(config *testConfig) int32 {
-			"scheduler-perf-",
-		)
+		// Mutate Node.
+		nodeAffinity.mutateNode(mutatedNodeTemplate)
```
lets delete `// Mutate Node.` and `// Mutate Pod.` comments , this code is self explanatory..

ravisantoshgudimetla replied:
ack.

jayunit100 commented:
/approve

jayunit100 commented:
@k8s-bot test this please

jayunit100 commented:
@k8s-bot ok to test.

jayunit100 commented:
Mostly superficial feedback.  IMO  a release note is not really  needed here.

sjug commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -203,6 +116,8 @@ func schedulePods(config *testConfig) int32 {
 		// TODO Find out why sometimes there may be scheduling blips in the beggining.
 		if len(scheduled) > config.numPods/100 {
+			//fmt.Println(config.numPods/100)
+			//time.Sleep(1 * time.Second)
```
What are these for?

sjug commented on test/utils/runners.go:
```diff
@@ -980,7 +980,7 @@ func makeCreatePod(client clientset.Interface, namespace string, podTemplate *v1
 }
 
-func createPod(client clientset.Interface, namespace string, podCount int, podTemplate *v1.Pod) error {
+func CreatePod(client clientset.Interface, namespace string, podCount int, podTemplate *v1.Pod) error {
```
Does this need to be exported?

ravisantoshgudimetla replied:
Yes, its being used in generatePods code.

ravisantoshgudimetla commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -240,107 +155,133 @@ func schedulePods(config *testConfig) int32 {
+func (inputConfig *schedulerPerfConfig) generatePods(config *testConfig) {
+	testutils.CreatePod(config.schedulerSupportFunctions.GetClient(), "sample", inputConfig.PodCount, config.mutatedPodTemplate)
+	testutils.CreatePod(config.schedulerSupportFunctions.GetClient(), "sample", config.numPods-inputConfig.PodCount, getBasePodTemplate())
 }
```
@sjug : This is where I am calling the CreatePod function.

ravisantoshgudimetla commented on test/integration/scheduler_perf/scheduler_perf_types.go:
```diff
@@ -18,11 +18,14 @@ package benchmark
 type nodeAffinity struct {
-	numGroups       int    // number of Node-Pod sets with Pods NodeAffinity matching given Nodes.
 	nodeAffinityKey string // Node Selection Key.
+	LabelCount      int    // number of labels to be added to each node or pod.
```
ack.

ravisantoshgudimetla replied:
I am not able to find the documentation for that. I will check and update @jayunit100

jayunit100 commented:
@ravisantoshgudimetla update bazel and rebase + push plz

sjug commented:
I'm not a fan of the functions that just return constants. Also, the function signatures don't match Cluster Loader patterns.
Otherwise it looks fine.

jayunit100 commented:
@k8s-bot test this please.

gmarek commented:
Is this targeted for 1.7?

ravisantoshgudimetla commented:
@sjug, @jayunit100 I have addressed your comments regarding fn's vs constants.(Here I made them as vars).

ravisantoshgudimetla commented:
> Is this targeted for 1.7?

I think so, @gmarek as this is part of 1.7 scheduling forensics

gmarek commented:
@jayunit100 @ravisantoshgudimetla - what do you plan to get from pushing it in 1.7, that you can't get from pushing it just after 1.8 opens?

jayunit100 commented:
I'm neutral on the 1.7 / 1.8 thing - but in general as ravi mentioned, yes, we had a goal of making scheduler perf as modular as possible for part of the scheduler forensics initiative.  This improvement is another step in that direction .  

Either way is ok with us, but obviously we'd like to keep building on it so 1.7 is great if it doesn't suck time from other things.

gmarek commented:
The thing is I'm out of review capacity for 1.7, so if you want my input it'll need to wait to until after 1.7 is cut. W can iterate after that and get it ready to merge just after 1.8 is open.

jayunit100 commented:
@k8s-bot ok to test

ravisantoshgudimetla commented:
> what do you plan to get from pushing it in 1.7, that you can't get from pushing it just after 1.8 opens?

As @jayunit100 suggested, it would great, if we can do it during 1.7 cycle as we intend to add few more predicates and priorities and see how they play together. I see that this could be useful for 
- isolated scheduler scale testing.
- Arbitrary combinations of predicate and priority.

We were planning to push the changes to core algorithm in 1.8(if any) based on the findings that we get from these changes.

> The thing is I'm out of review capacity for 1.7, so if you want my input it'll need to wait to until after 1.7 is cut. W can iterate after that and get it ready to merge just after 1.8 is open.

We are fine with pushing these changes to 1.8 too.

gmarek commented:
I don't see any benefit of pushing this into 1.7. I believe that those changes are isolated from other work that's going on, so rebases are not an issue, and you can easily run tests on patched branch. I'd rather do the review in the post-1.7 timeframe, when there'll be no rush and we'll have time for proper review.

ravisantoshgudimetla commented:
> I don't see any benefit of pushing this into 1.7. I believe that those changes are isolated from other work that's going on, so rebases are not an issue, and you can easily run tests on patched branch. I'd rather do the review in the post-1.7 timeframe, when there'll be no rush and we'll have time for proper review.

Sure @gmarek, we can wait.

aveshagarwal commented:
@kubernetes/sig-scheduling-pr-reviews

ravisantoshgudimetla commented:
/cc @k82cn, I think you wanted to do something similar with benchmarking for individual predicates and priorities, which is a subset of these changes. Please let me know your feedback.

k82cn commented:
/assign 

assign to myself, there's a requirement to get run some profiling for each priority/predicate to see 'who' consuming times; that'll be better if sched_perf can handle it.

jayunit100 commented:
Thanks k82cn.  I know ravi also was or is working on this, please coordinate .  The PRs for improving the architecture of sched_perf  I believe go in 1.8, but can be used now to test easily just pulling the bits down into a branch. Cc me on issues I'll try to follow along and help with revues as well.

k82cn commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -54,82 +89,6 @@ func TestSchedule100Node3KPods(t *testing.T) {
 }
 
-// TestSchedule100Node3KNodeAffinityPods schedules 3k pods using Node affinity on 100 nodes.
-func TestSchedule100Node3KNodeAffinityPods(t *testing.T) {
```
Why remove those test?

ravisantoshgudimetla replied:
This can be simulated by extending the datastructure, so there is need to have a separate test function. For example, when we have affinity set in schedulerPerfConfig struct(https://github.com/kubernetes/kubernetes/pull/46117/files/4ba70af63f9793649e3cf5108d6ff5cd74858c9b#diff-3126019cd381f25186f4fa8c8da06ec6R267), this will be automatically run.

jayunit100 replied:
@ravisantoshgudimetla true .......... But I thought we are **keeping the test**, and just make it a thin wrapper?  We want test results to be identical ...

jayunit100 commented:
I'd like  to figure out who is going to be maintaining this work also so we can have some context to the review process, there was a roadmap for this work.

@ravisantoshgudimetla @sjug @jeremyeder  are you interested/capable in maintaining this going forward ?

ravisantoshgudimetla commented:
I am interested in maintaining this @jayunit100.

jayunit100 commented:
OK @ravisantoshgudimetla , So the long term goal here is the kitchen sink, right?  If so, lets make a new PR branching off this that shows how to implement that, maybe even a lightning talk for sig-scheduling.  The kitchen sink tests where we can simulate different pred/prio situations is the main justification for this work as i recall.

ravisantoshgudimetla committed:
# This is a combination of 2 commits.
# The first commit's message is:

Modular extensions for kube scheduler perf testing framework

# This is the 2nd commit message:

Modular extensions for kube scheduler perf testing framework

ravisantoshgudimetla committed:
Build files generated

ravisantoshgudimetla commented:
/retest

ravisantoshgudimetla commented:
/retest

ravisantoshgudimetla commented:
ping @gmarek for a review. @jayunit100 @k82cn Do you have any other comments?

gmarek commented:
I lost most of the context... Looks good to me now, and we can always iterate later if something pops up.

jayunit100 commented:
I had a few minor comments to address nothing major.  Longer term is that you guys are using this if we're going to keep refactoring, would love to see a README with examples of how to generate data and simulate things in a follow on PR.

ravisantoshgudimetla commented:
Ok can someone approve/lgtm then?

gmarek commented:
/lgtm

gmarek commented:
/approve

jayunit100 commented:
My comments can be addressed in a follow on.

Ravi are you going to do that or is someone else ?

ravisantoshgudimetla commented:
>  Also test infra -> CI running ...
For that, as you know, I prepared the slides but have datapoints for taints & tolerations but as suggested we need more predicates or priorities. I may get started on it next week. 

>  Ravi are you going to do that or is someone else ?
Hopefully me but I am not sure, if you have some concrete plan for 1.8

ravisantoshgudimetla commented:
/retest
